File: src/main/java/rars/venus/editors/TokenStyle.java
package rars.venus.editors;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;

public record TokenStyle(
        @Nullable Color foreground,
        @Nullable Color background,
        boolean isBold,
        boolean isItalic,
        boolean isUnderline
) {
    public static final Color DEFAULT_FOREGROUND = null;
    public static final Color DEFAULT_BACKGROUND = null;

    public static final TokenStyle DEFAULT = new TokenStyle(DEFAULT_FOREGROUND, DEFAULT_BACKGROUND, false, false,
            false);

    public static @NotNull TokenStyle plain(final @NotNull Color foreground) {
        return new TokenStyle(foreground, DEFAULT_BACKGROUND, false, false, false);
    }

    public static @NotNull TokenStyle bold(final @NotNull Color foreground) {
        return new TokenStyle(foreground, DEFAULT_BACKGROUND, true, false, false);
    }

    public static @NotNull TokenStyle italic(final @NotNull Color foreground) {
        return new TokenStyle(foreground, DEFAULT_BACKGROUND, false, true, false);
    }
    public static @NotNull TokenStyle underline(final @NotNull Color foreground) {
        return new TokenStyle(foreground, DEFAULT_BACKGROUND, false, false, true);
    }
}

File: src/main/java/rars/venus/editors/TokenStyle.kt
package rars.venus.editors

import java.awt.Color

@JvmRecord
data class TokenStyle(
    val foreground: Color?,
    val background: Color?,
    val isBold: Boolean,
    val isItalic: Boolean,
    val isUnderline: Boolean
) : Cloneable {
    override fun clone(): TokenStyle = super.clone() as TokenStyle

    companion object {
        val DEFAULT_FOREGROUND: Color? = null
        val DEFAULT_BACKGROUND: Color? = null

        @JvmField
        val DEFAULT: TokenStyle = TokenStyle(
            DEFAULT_FOREGROUND, DEFAULT_BACKGROUND,
            isBold = false, isItalic = false, isUnderline = false
        )

        fun plain(foreground: Color): TokenStyle = TokenStyle(
            foreground, DEFAULT_BACKGROUND,
            isBold = false, isItalic = false, isUnderline = false
        )

        fun bold(foreground: Color): TokenStyle = TokenStyle(
            foreground, DEFAULT_BACKGROUND,
            isBold = true, isItalic = false, isUnderline = false
        )

        fun italic(foreground: Color): TokenStyle = TokenStyle(
            foreground, DEFAULT_BACKGROUND,
            isBold = false, isItalic = true, isUnderline = false
        )

        fun underline(foreground: Color): TokenStyle = TokenStyle(
            foreground, DEFAULT_BACKGROUND,
            isBold = false, isItalic = false, isUnderline = true
        )
    }
}
